README.md
RM-Synthesis is technique used to transform polarized intensity between wavelength-space (lambda) and faraday-space (RM). See Brentjen & de Bruyn (2005) for more detail. 

This technqiue can become computationally expensive. For example, wide-field views imply a large number of line-of-sights (pixels) to evaluate, and wide spectral-bandwidth, although necessary, imply a large number of channels to average across. Thus, optimizing RM-Synthesis technqiue is important. Particularly for all-sky surveys such as VLASS, whereby all pixels need to be evaluated -- this is unlike the case where specific pixel locations are known and of interest.
In the original implementation, all pixels within lambda^2-plane are evaluated at once, although this is may have been optimum at that time, the large dataset of modern instruments makes it infeasible.

In our RM-Synthesis code, we evaluate each pixel at a time. This approach as is is the less optimum thus, we have incorporated multiprocessing (and we intend to extend this to GPUs) so that multiple pixels can we evaluated at once. This was tested on Cyngus A 2k by 2k images with 2-18 GHz bandwidth (~1000 channels). Without parallel processing, this task takes over > 20 hours to run and with multi-processing using 6 cores this takes exactly 2 hours. When using the original approach, we encounter Memory Error since the combined data size of our Stokes Q and U is 33 GB. 

1. Installation:







2. Executing:



CHANGES.md
# 0.1.2


Moved from version 0.1.1 to 0.1.2
Computing everything within the pool and not outside. 


rm_synthesis.py
#!/usr/bin/env python
## Lerato Sebokolodi <mll.sebokolodi@gmail.com>

import numpy
import astropy.io.fits as pyfits
import time
import sys
from multiprocessing import Pool
import argparse




def read_data(image, freq=True):
                                                                              
    """ Read and return image data: ra, dec and freq only"""
    try:
        with pyfits.open(image) as hdu:
            imagedata = hdu[0].data
            header = hdu[0].header
        imslice = numpy.zeros(imagedata.ndim, dtype=int).tolist()
        imslice[-1] = slice(None)
        imslice[-2] = slice(None)
        if freq:
            imslice[-3] = slice(None)
        print('>>> Image %s loaded sucessfully. '%image)
        return imagedata[imslice], header
    except OSError:
        sys.exit('>>> could not load %s. Aborting...'%image)
        

def check_shape(qfits, ufits, frequencies):
    """
    Checks the shape of the cubes and frequency file
    """
    qhdr =  pyfits.getheader(qfits)
    uhdr =  pyfits.getheader(ufits)
    
    errors = []
    axis = ['naxis1', 'naxis2', 'naxis3']
    if qhdr['naxis'] < 3 or uhdr['naxis'] < 3:
        errors.append('The dimensions of Q = %d and U = %d, not >=3.' %(
               qhdr['naxis'], uhdr['naxis']))
    if qhdr['naxis'] != uhdr['naxis']:
        if qhdr['naxis'] >= 3 and uhdr['naxis'] >=3:
             for ax in axis:
                 if qhdr[ax] != uhdr[ax]:
                      errors.append('%s for Q is != %d of U.' %(ax, qhdr[ax], uhdr[ax]))
    if qhdr[axis[2]] != len(frequencies) or uhdr[axis[2]] != len(frequencies):
        errors.append('Freq-axis of the cubes differ from that of the '
           'frequency file provided.')
            
    return errors


def create_mask(image):
    """
    This creates a mask.
    """

    return 


def faraday_phase(phi_sample, wavelengths):       
    
    """return phase term of the Faraday spectrum """
    # ra, dec, wavelengths, faraday depth
    phase = numpy.zeros([len(phi_sample), len(wavelengths)], dtype=numpy.complex64)  
    derotated_wavelength = wavelengths - wavelengths.mean()
    for i, phi in enumerate(phi_sample):                  
        phase[i, :] = numpy.exp(-2 * 1j * phi * derotated_wavelength )
                              
    return phase



def add_RM_to_fits_header(header, phi_sample=None):

    """
    Adds RM axis to the data in place of frequency.

    NB:Knicked and modified from Brentjens RM Synthesis code     
    """
    hdr = header.copy()
    if len(phi_sample) >= 2:
        # if it is a cube then do this:
        new_hdr = {'naxis3': len(phi_sample), 'crpix3':1.0,
                  'cdelt3': phi_sample[1]-phi_sample[0], 
                  'crval3': phi_sample[0], 'ctype3':'RM',
                  'cunit3': 'rad/m^2'}
    else:
        new_hdr = {'naxis': 3, 'naxis3': 1, 'cunit3': 'rad/m^2', 
                   'ctype3':'RM'}

    hdr.update(new_hdr) 
    return hdr


def faraday_depth_interval(wavelengths, phi_max=None,  
                       phi_min=None, dphi=None):

    """Computes the RM interval and resolution"""
    if phi_max is None:
        phi_max = (1.9/abs(numpy.diff(wavelengths)).max())

    if phi_min is None:
       phi_min = -phi_max

    if dphi is None:
       dphi = (3.8/abs(wavelengths[0] - wavelengths[-1]))/3.0

    return phi_max, phi_min, dphi



def compute_dispersion(x, y):
   
    """Computes Faraday Dispersion of a pixel in question"""

    start = time.time()
    P = qdata[:, x, y] + 1j * udata[:, x, y]
    dispersion = numpy.sum(P * phase, axis=1)/N_wave
    end = time.time()
    print('Time %.6f secs for (%d, %d)'%(end-start, x, y))
    return dispersion





if __name__=='__main__':

    parser = argparse.ArgumentParser(description='Performs 1-D RM Synthesis. '
             'It requires Stokes Q and U image cubes as well as a frequency '
             'in text format. These cubes should be (312), freq, ra, dec. '
             'There is an option for multi-processing, see numProcessor. '
             'The outputs are the Q and U  of Faraday dispersion (FD) cube, '
             'RM map derived from peak in FD spectrum. A png plot of the RMSF. ')
    add = parser.add_argument
    add('-q', '--qcube', dest='qfits', help='Stokes Q cube (fits)')
    add('-u', '--ucube', dest='ufits', help='Stokes U cube (fits)')
    add('-f', '--freq', dest='freq', help='Frequency file (text)')  
    add('-np', '--numpr', dest='numProcessor', help='number of cores to use. Default 1.', 
        default=1, type=int)
    add('-rn', '-rm-min', dest='phi_min', help='Minimum Faraday depth. Default None. '
       'If not specified will be determined internally. See Brentjens (2005) paper.',
       type=float, default=None)
    add('-rx', '-rm-max', dest='phi_max', help='Maximum Faraday depth. See -rn.',
       type=float, default=None)
    add('-rs', '-rm-sample', dest='dphi', help='The sampling width in Faraday-space.',
        type=float, default=None)
    add('-o', '--prefix', dest='prefix', help='This is a prefix for output files.')
    
    
    #TODO: 1) RM cleaning, '
    #TODO: 2) derotated the observed PA by RM_peak *wavelengths^2
    #TODO: 3) take mask
    #TODO: 4) add log file.
    #TODO: 5) turn this into a software

    args = parser.parse_args()

    try:
        frequencies = numpy.loadtxt(args.freq)
    except ValueError:
        sys.exit(">>> Problem found with frequency file. It should be a text file")
    
    
    errors = check_shape(args.qfits, args.ufits, frequencies)
    if len(errors) > 0:
        print(errors)
        sys.exit()
  
    wavelengths =  (299792458.0/frequencies)**2 
    qdata, qhdr = read_data(args.qfits) # run Q-cube 
    udata, uhdr = read_data(args.ufits) # run U-cube

    
    phi_max, phi_min, dphi = faraday_depth_interval(wavelengths, args.phi_max, 
                         args.phi_min, args.dphi)
    phi_sample =  numpy.arange(phi_min, phi_max, dphi)
    print('>>> Maximum RM = %.2f, minimum RM =%.2f, in steps of %.2f rad/m^2'\
          %(phi_max, phi_min, dphi))

    phase = faraday_phase(phi_sample, wavelengths)
    N_wave, N_x, N_y = qdata.shape
    N_phi = len(phi_sample)
    

    x, y = numpy.indices((N_x, N_y))
    x = x.flatten()
    y = y.flatten()
    Faraday_Dispersion = numpy.zeros([N_phi, N_x, N_y ], dtype=numpy.complex64)
    start_all= time.time()
    pool = Pool(args.numProcessor)  
    for (xx, yy) in zip(x, y):  
       Faraday_Dispersion[:, xx, yy] = pool.apply(compute_dispersion, args=(xx, yy))
    end_all = time.time()
    print('The total time for multiprocessing is %.6f'%(end_all-start_all))
    Faraday_amp = numpy.absolute(Faraday_Dispersion)
    peaks = numpy.argmax(Faraday_amp, axis=0)
    RM_peak = phi_sample[peaks]
    rmhdr = add_RM_to_fits_header(qhdr, phi_sample=[1])
    args.prefix = args.prefix or args.qfits.split('.')[0]
    pyfits.writeto(args.prefix + '-RM.FITS', RM_peak, rmhdr, clobber=True)
    
    fdhdr = add_RM_to_fits_header(qhdr, phi_sample=phi_sample)
    pyfits.writeto(args.prefix + '-QDISPER.FITS', Faraday_Dispersion.real, fdhdr, clobber=True)
    pyfits.writeto(args.prefix + '-UDISPER.FITS', Faraday_Dispersion.imag, fdhdr, clobber=True)
    
    
    



RM-Synthesis-1D.py
#!/usr/bin/env python
## Lerato Sebokolodi <mll.sebokolodi@gmail.com>

import numpy
import pyfits
import time
import sys
from multiprocessing import Pool
import argparse




def read_data(image, freq=True):
                                                                              
    """ Read and return image data: ra, dec and freq only"""
    try:
        with pyfits.open(image) as hdu:
            imagedata = hdu[0].data
            header = hdu[0].header
        imslice = numpy.zeros(imagedata.ndim, dtype=int).tolist()
        imslice[-1] = slice(None)
        imslice[-2] = slice(None)
        if freq:
            imslice[-3] = slice(None)
        print('>>> Image %s loaded sucessfully. '%image)
        return imagedata[imslice], header
    except OSError:
        sys.exit('>>> could not load %s. Aborting...'%image)
        

def check_shape(qfits, ufits, frequencies):
    """
    Checks the shape of the cubes and frequency file
    """
    qhdr =  pyfits.getheader(qfits)
    uhdr =  pyfits.getheader(ufits)
    
    errors = []
    axis = ['naxis1', 'naxis2', 'naxis3']
    if qhdr['naxis'] < 3 or uhdr['naxis'] < 3:
        errors.append('The dimensions of Q = %d and U = %d, not >=3.' %(
               qhdr['naxis'], uhdr['naxis']))
    if qhdr['naxis'] != uhdr['naxis']:
        if qhdr['naxis'] >= 3 and uhdr['naxis'] >=3:
             for ax in axis:
                 if qhdr[ax] != uhdr[ax]:
                      errors.append('%s for Q is != %d of U.' %(ax, qhdr[ax], uhdr[ax]))
    if qhdr[axis[2]] != len(frequencies) or uhdr[axis[2]] != len(frequencies):
        errors.append('Freq-axis of the cubes differ from that of the '
           'frequency file provided.')
            
    return errors


def create_mask(image):
    """
    This creates a mask.
    """

    return 


def faraday_phase(phi_sample, wavelengths):       
    
    """return phase term of the Faraday spectrum """
    # ra, dec, wavelengths, faraday depth
    phase = numpy.zeros([len(phi_sample), len(wavelengths)], dtype=numpy.complex64)  
    derotated_wavelength = wavelengths - wavelengths.mean()
    for i, phi in enumerate(phi_sample):                  
        phase[i, :] = numpy.exp(-2 * 1j * phi * derotated_wavelength )
                              
    return phase



def add_RM_to_fits_header(header, phi_sample=None):

    """
    Adds RM axis to the data in place of frequency.

    NB:Knicked and modified from Brentjens RM Synthesis code     
    """
    hdr = header.copy()
    if len(phi_sample) >= 2:
        # if it is a cube then do this:
        new_hdr = {'naxis3': len(phi_sample), 'crpix3':1.0,
                  'cdelt3': phi_sample[1]-phi_sample[0], 
                  'crval3': phi_sample[0], 'ctype3':'RM',
                  'cunit3': 'rad/m^2'}
    else:
        new_hdr = {'naxis': 3, 'naxis3': 1, 'cunit3': 'rad/m^2', 
                   'ctype3':'RM'}

    hdr.update(new_hdr) 
    return hdr


def faraday_depth_interval(wavelengths, phi_max=None,  
                       phi_min=None, dphi=None):

    """Computes the RM interval and resolution"""
    if phi_max is None:
        phi_max = (1.9/abs(numpy.diff(wavelengths)).max())

    if phi_min is None:
       phi_min = -phi_max

    if dphi is None:
       dphi = (3.8/abs(wavelengths[0] - wavelengths[-1]))/3.0

    return phi_max, phi_min, dphi



def compute_dispersion(x, y):
   
    """Computes Faraday Dispersion of a pixel in question"""

    start = time.time()
    P = qdata[:, x, y] + 1j * udata[:, x, y]
    dispersion = numpy.sum(P * phase, axis=1)/N_wave
    end = time.time()
    print('Time %.6f secs for (%d, %d)'%(end-start, x, y))
    return dispersion





if __name__=='__main__':

    parser = argparse.ArgumentParser(description='Performs 1-D RM Synthesis. '
             'It requires Stokes Q and U image cubes as well as a frequency '
             'in text format. These cubes should be (312), freq, ra, dec. '
             'There is an option for multi-processing, see numProcessor. '
             'The outputs are the Q and U  of Faraday dispersion (FD) cube, '
             'RM map derived from peak in FD spectrum. A png plot of the RMSF. ')
    add = parser.add_argument
    add('-q', '--qcube', dest='qfits', help='Stokes Q cube (fits)')
    add('-u', '--ucube', dest='ufits', help='Stokes U cube (fits)')
    add('-f', '--freq', dest='freq', help='Frequency file (text)')  
    add('-np', '--numpr', dest='numProcessor', help='number of cores to use. Default 1.', 
        default=1, type=int)
    add('-rn', '-rm-min', dest='phi_min', help='Minimum Faraday depth. Default None. '
       'If not specified will be determined internally. See Brentjens (2005) paper.',
       type=float, default=None)
    add('-rx', '-rm-max', dest='phi_max', help='Maximum Faraday depth. See -rn.',
       type=float, default=None)
    add('-rs', '-rm-sample', dest='dphi', help='The sampling width in Faraday-space.',
        type=float, default=None)
    add('-o', '--prefix', dest='prefix', help='This is a prefix for output files.')
    
    
    #TODO: 1) RM cleaning, '
    #TODO: 2) derotated the observed PA by RM_peak *wavelengths^2
    #TODO: 3) take mask
    #TODO: 4) add log file.
    #TODO: 5) turn this into a software

    args = parser.parse_args()

    try:
        frequencies = numpy.loadtxt(args.freq)
    except ValueError:
        sys.exit(">>> Problem found with frequency file. It should be a text file")
    
    
    errors = check_shape(args.qfits, args.ufits, frequencies)
    if len(errors) > 0:
        print(errors)
        sys.exit()
  
    wavelengths =  (299792458.0/frequencies)**2 
    qdata, qhdr = read_data(args.qfits) # run Q-cube 
    udata, uhdr = read_data(args.ufits) # run U-cube

    
    phi_max, phi_min, dphi = faraday_depth_interval(wavelengths, args.phi_max, 
                         args.phi_min, args.dphi)
    phi_sample =  numpy.arange(phi_min, phi_max, dphi)
    print('>>> Maximum RM = %.2f, minimum RM =%.2f, in steps of %.2f rad/m^2'\
          %(phi_max, phi_min, dphi))

    phase = faraday_phase(phi_sample, wavelengths)
    N_wave, N_x, N_y = qdata.shape
    N_phi = len(phi_sample)
    

    x, y = numpy.indices((N_x, N_y))
    x = x.flatten()
    y = y.flatten()
    Faraday_Dispersion = numpy.zeros([N_phi, N_x, N_y ], dtype=numpy.complex64)

    pool = Pool(args.numProcessor)  
    for (xx, yy) in zip(x, y):  
       Faraday_Dispersion[:, xx, yy] = pool.apply(compute_dispersion, args=(xx, yy))
   
    Faraday_amp = numpy.absolute(Faraday_Dispersion)
    peaks = numpy.argmax(Faraday_amp, axis=0)
    RM_peak = phi_sample[peaks]
    rmhdr = add_RM_to_fits_header(qhdr, phi_sample=[1])
    args.prefix = args.prefix or args.qfits.split('.')[0]
    pyfits.writeto(args.prefix + '-RM.FITS', RM_peak, rmhdr, overwrite=True)
    
    fdhdr = add_RM_to_fits_header(qhdr, phi_sample=phi_sample)
    pyfits.writeto(args.prefix + '-QDISPER.FITS', Faraday_Dispersion.real, fdhdr, overwrite=True)
    pyfits.writeto(args.prefix + '-UDISPER.FITS', Faraday_Dispersion.imag, fdhdr, overwrite=True)
    



rmsynthesis/rmsynthesis.py
## Lerato Sebokolodi <mll.sebokolodi@gmail.com>

import numpy
import astropy.io.fits as pyfits
import time
import sys
from multiprocessing import Pool
import argparse
import os


_path = os.path.realpath(__file__)
_path = os.path.dirname(_path)

import pkg_resources
try:
    __version__ = pkg_resources.require("rm-synthesis")[0].version
except pkg_resources.DistributionNotFound:
    __version__ = "dev"


def read_data(image, freq=True):
                                                                              
    """ Read and return image data: ra, dec and freq only"""
    try:
        with pyfits.open(image) as hdu:
            imagedata = hdu[0].data
            header = hdu[0].header
        imslice = numpy.zeros(imagedata.ndim, dtype=int).tolist()
        imslice[-1] = slice(None)
        imslice[-2] = slice(None)
        if freq:
            imslice[-3] = slice(None)
        print('>>> Image %s loaded sucessfully. '%image)
        return imagedata[imslice], header
    except OSError:
        sys.exit('>>> could not load %s. Aborting...'%image)
        


def check_shape(qfits, ufits, frequencies):
    """
    Checks the shape of the cubes and frequency file
    """
    qhdr =  pyfits.getheader(qfits)
    uhdr =  pyfits.getheader(ufits)
    
    errors = []
    axis = ['naxis1', 'naxis2', 'naxis3']
    if qhdr['naxis'] < 3 or uhdr['naxis'] < 3:
        errors.append('The dimensions of Q = %d and U = %d, not >=3.' %(
               qhdr['naxis'], uhdr['naxis']))
    if qhdr['naxis'] != uhdr['naxis']:
        if qhdr['naxis'] >= 3 and uhdr['naxis'] >=3:
             for ax in axis:
                 if qhdr[ax] != uhdr[ax]:
                      errors.append('%s for Q is != %d of U.' %(ax, qhdr[ax], uhdr[ax]))
    if qhdr[axis[2]] != len(frequencies) or uhdr[axis[2]] != len(frequencies):
        errors.append('Freq-axis of the cubes differ from that of the '
           'frequency file provided.')
            
    return errors

 

def faraday_phase(phi_sample, wavelengths):       
    
    """return phase term of the Faraday spectrum """
    phase = numpy.zeros([len(phi_sample), len(wavelengths)], dtype=numpy.complex64)  
    derotated_wavelength = wavelengths - wavelengths.mean()
    for i, phi in enumerate(phi_sample):                  
        phase[i, :] = numpy.exp(-2 * 1j * phi * derotated_wavelength )
                              
    return phase



def add_RM_to_fits_header(header, pol, phi_sample=None):

    """
    Adds RM axis to the data in place of frequency.

    NB:Knicked and modified from Brentjens RM Synthesis code     
    """
    hdr = header.copy()
    if len(phi_sample) >= 2:
        # if it is a cube then do this:
        new_hdr = {'naxis3': len(phi_sample),
                   'crpix3':1.0, 'cdelt3': phi_sample[1]-phi_sample[0], 
                   'crval3': phi_sample[0], 'ctype3':'RM',
                   'cunit3': 'rad/m^2', 'POL': '%s'%pol}
                  
    else:
        new_hdr = {'naxis3': 1, 'cunit3': 'rad/m^2', 
                   'ctype3':'RM', 'POL': '%s'%pol}

    hdr.update(new_hdr) 
    return hdr


def faraday_depth_interval(wavelengths, phi_max=None,  
                       phi_min=None, dphi=None):

    """Computes the RM interval and resolution"""
    #TODO: optimize this better booboo
    if phi_max is None:
        phi_max = (1.9/abs(numpy.diff(wavelengths)).max())

    if phi_min is None:
       phi_min = -phi_max

    if dphi is None:
       dphi = (3.8/abs(wavelengths[0] - wavelengths[-1]))/3.0

    return phi_max, phi_min, dphi



def compute_dispersion(x, y):
   
    """Computes Faraday Dispersion of a pixel in question"""

    start = time.time()
    P = qdata[:, x, y] + 1j * udata[:, x, y]
    dispersion = numpy.sum(P * phase, axis=1)/N_wave
    dispersion_amp = numpy.absolute(dispersion)
    ind = numpy.where(dispersion_amp == max(dispersion_amp))[0]
    end = time.time()
    print('Time %.6f secs for (%d, %d)'%(end-start, x, y))
    return dispersion, dispersion_amp, ind


def read_mask(fits):
    """get pixel coordinates for a mask"""
    maskdata, mhdr = read_data(fits, freq=False)
    xpix, ypix = numpy.where(maskdata > 0)
    return xpix, ypix



def save_fits(prefix, tag, header, sample, tosave):

    """save fits files"""
    hdr = add_RM_to_fits_header(header, tag, sample)
    pyfits.writeto(prefix + '-%s.FITS'%tag, tosave, hdr, 
           overwrite=True)


def main():

    parser = argparse.ArgumentParser(description='Performs 1-D RM Synthesis. '
             'It requires Stokes Q and U image cubes as well as a frequency '
             'in text format. These cubes should be (312), freq, ra, dec. '
             'There is an option for multi-processing, see numProcessor. '
             'The outputs are the Q and U  of Faraday dispersion (FD) cube, '
             'RM map derived from peak in FD spectrum. The RMSF amplitude and '
             'the RM range in a text file.')
    add = parser.add_argument
    add("-v","--version", action="version",version=
        "{:s} version {:s}".format(parser.prog, __version__))
    add('-q', '--qcube', dest='qfits', help='Stokes Q cube (fits)')
    add('-u', '--ucube', dest='ufits', help='Stokes U cube (fits)')
    add('-f', '--freq', dest='freq', help='Frequency file (text)')  
    add('-ncore', '--numpr', dest='numProcessor', help='number of cores to use. Default 1.', 
        default=1, type=int)
    add('-rmin', '--rm-min', dest='phi_min', help='Minimum Faraday depth. Default None. '
       'If not specified will be determined internally. See Brentjens (2005) paper.',
       type=float, default=None)
    add('-rmax', '--rm-max', dest='phi_max', help='Maximum Faraday depth. See -rn.',
       type=float, default=None)
    add('-rstep', '--rm-sample', dest='dphi', help='The sampling width in Faraday-space.',
        type=float, default=None)
    add('-mask', '--maskfits', dest='maskfits', help='A mask image (fits). This package '
        'comes with a tool called cleanmask to allow the user to create a mask, for more info '
        'cleanmask -h', default=None)
    add('-o', '--prefix', dest='prefix', help='This is a prefix for output files.')
    
    
    #TODO: 1) RM cleaning, '
    #TODO: 2) derotate the observed PA by RM_peak *wavelengths^2
    #TODO: 4) add log file.

    args = parser.parse_args()

    try:
        frequencies = numpy.loadtxt(args.freq)
    except ValueError:
        sys.exit(">>> Problem found with frequency file. It should be a text file")
    
    
    errors = check_shape(args.qfits, args.ufits, frequencies)
    if len(errors) > 0:
        print(errors)
        sys.exit()
    global qdata, udata, phase, N_wave

    wavelengths =  (299792458.0/frequencies)**2 
    qdata, qhdr = read_data(args.qfits) # run Q-cube 
    udata, uhdr = read_data(args.ufits) # run U-cube

    
    phi_max, phi_min, dphi = faraday_depth_interval(wavelengths, args.phi_max, 
                         args.phi_min, args.dphi)
    phi_sample =  numpy.arange(phi_min, phi_max, dphi)
    print('>>> Maximum RM = %.2f, minimum RM =%.2f, in steps of %.2f rad/m^2'\
          %(phi_max, phi_min, dphi))

    phase = faraday_phase(phi_sample, wavelengths)
    N_wave, N_x, N_y = qdata.shape
    N_phi = len(phi_sample)
    
    # use mask or all pixels
    if args.maskfits:
        x, y = read_mask(args.maskfits)
    
    else:
        x, y = numpy.indices((N_x, N_y))
        x = x.flatten()
        y = y.flatten()
    
    wave_mean = wavelengths.mean()
    # compute the Faraday dispersion via multi-processing
    Faraday_Dispersion = numpy.zeros([N_phi, N_x, N_y ], dtype=numpy.complex64)
    Amp_Dispersion = numpy.zeros([N_x, N_y])
    RM_Peak = numpy.zeros([N_x, N_y])
    Angle_Peak = numpy.zeros([N_x, N_y])

    start_all = time.time()
    pool = Pool(args.numProcessor)
    for i, (xx, yy) in enumerate(zip(x, y)):  
       dispersion, amp, ind = pool.apply(compute_dispersion, args=(xx, yy))
       Faraday_Dispersion[:, xx, yy]  = dispersion
       Amp_Dispersion[xx, yy] = amp[ind]
       RM = phi_sample[ind] 
       RM_Peak[xx, yy] = RM
       Angle_Peak[xx, yy] = numpy.angle(dispersion[ind]) - RM * wave_mean
    end_all = time.time()
    print('>>> The total time for multiprocessing is %.6f.'%(end_all-start_all))
  
    phase_amp = numpy.absolute(numpy.sum(phase, axis=1) )/N_wave
    print('>>> Saving outputs...')
    args.prefix = args.prefix or args.qfits.split('.')[0]
    numpy.savetxt(args.prefix  + '-RMSF.txt', numpy.vstack((phi_sample, phase_amp)) )
    save_fits(args.prefix, 'RM' , qhdr, [1], RM_Peak)
    save_fits(args.prefix, 'AMP', qhdr, [1], Amp_Dispersion)
    save_fits(args.prefix, 'ANG', qhdr, [1], Angle_Peak)
    save_fits(args.prefix, 'QFAR', qhdr, phi_sample, Faraday_Dispersion.real)
    save_fits(args.prefix, 'UFAR', qhdr, phi_sample, Faraday_Dispersion.imag)
    print('>>> Done!')


